File: Army/MagicAttack.cpp
#include "MagicAttack.hpp"
#include "SpellCaster.hpp"
#include "Unit.hpp"

MagicAttack::MagicAttack(int magicPower, int manaConsume): Spell(magicPower, manaConsume) {}
MagicAttack::~MagicAttack() {}

void MagicAttack::spellCast(Unit& enemy, SpellCaster& thisUnit) {

    if ( thisUnit.getMana().getCurrentMana() >= this->manaConsume ) {
        int potentialPower = this->magicPower;

        if ( thisUnit.getState().isHealer() ) {
            potentialPower /= 2;
        }
        if ( enemy.getState().isCanTakeMagicDamage() ) {
            enemy.getHealth().takeDamage(potentialPower);
        }
        thisUnit.getMana().takeMana(this->manaConsume);
    }
}

File: Army(java)/MagicAttack.java
public class MagicAttack extends Spell {
    public MagicAttack(int magicPower, int manaConsume) {
        super(magicPower, manaConsume);
    }

    @Override
    public void affect(Unit enemy, int power) {
        enemy.getHealth().takeDamage(power);
    }
}